crux_cli/src/codegen/mod.rs

```diff
@@ -0,0 +1,160 @@
+mod filter;
+mod formatter;
+pub mod generate;
+mod indexed;
+mod item;
+mod node;
+mod serde;
+mod serde_generate;
+
+use std::{
+    collections::{BTreeMap, HashMap},
+    fs::{self, File},
+    io::Read,
+    path::PathBuf,
+    process::Command,
+};
+
+use anyhow::{anyhow, bail, Context, Result};
+use guppy::{graph::PackageGraph, MetadataCommand};
+use log::debug;
+use rustdoc_types::Crate;
+
+use crate::args::CodegenArgs;
+use filter::Filter;
+use formatter::Formatter;
+use node::ItemNode;
+use serde_generate::format::ContainerFormat;
+
+pub type Registry = BTreeMap<String, ContainerFormat>;
+
+pub fn codegen(args: &CodegenArgs) -> Result<()> {
+    let mut cmd = MetadataCommand::new();
+    let package_graph = PackageGraph::from_command(&mut cmd)?;
+
+    let manifest_paths: BTreeMap<&str, &str> = package_graph
+        .packages()
+        .map(|package| (package.name(), package.manifest_path().as_str()))
+        .collect();
+
+    let Ok(lib) = package_graph.workspace().member_by_path(&args.lib) else {
+        bail!("Could not find workspace package with path {}", args.lib)
+    };
+
+    let lib_name = lib.name();
+
+    let registry = run(lib_name, |name| load_crate(name, &manifest_paths))?;
+
+    // switch from vendored types to `serde-reflection` types
+    let registry: serde_reflection::Registry =
+        serde_json::from_slice(&serde_json::to_vec(&registry)?)?;
+
+    fs::create_dir_all(&args.output)?;
+
+    generate::java(&registry, &args.java_package, args.output.join("java"))
+        .context("Generating types for Java")?;
+
+    generate::swift(&registry, &args.swift_package, args.output.join("swift"))
+        .context("Generating tyeps for Swift")?;
+
+    generate::typescript(
+        &registry,
+        &args.typescript_package,
+        &lib.version().to_string(),
+        args.output.join("typescript"),
+    )
+    .context("Generating types for TypeScript")?;
+    Ok(())
+}
+
+fn run<F>(crate_name: &str, load: F) -> Result<Registry>
+where
+    F: Fn(&str) -> Result<Crate>,
+{
+    let mut previous: HashMap<String, Crate> = HashMap::new();
+
+    let shared_lib = load(crate_name)?;
+
+    let mut filter = Filter::default();
+    filter.process(crate_name, &shared_lib)?;
```

Comment:
These things need better names, IMO. 

It's not clear what sort of filtering is happening, for what purpose... Not a huge priority now, but, though I'd point it out.

Reply:
Agreed. We can refactor later I think.